contracts/PublicFundraising.sol

```diff
@@ -187,21 +182,81 @@ contract PublicFundraising is
         tokensSold += _amount;
         tokensBought[_tokenReceiver] += _amount;
 
-        // rounding up to the next whole number. Investor is charged up to one currency bit more in case of a fractional currency bit.
-        uint256 currencyAmount = Math.ceilDiv(_amount * getPrice(), 10 ** token.decimals());
+        token.mint(_tokenReceiver, _amount);
+    }
 
+    function _getFeeAndFeeReceiver(uint256 _currencyAmount) internal view returns (uint256, address) {
+        IFeeSettingsV2 feeSettings = token.feeSettings();
+        return (feeSettings.publicFundraisingFee(_currencyAmount), feeSettings.publicFundraisingFeeCollector());
+    }
+
+    /**
+     * @notice Buy `amount` tokens and mint them to `_tokenReceiver`.
+     * @param _amount amount of tokens to buy, in bits (smallest subunit of token)
+     * @param _tokenReceiver address the tokens should be minted to
+     */
+    function buy(uint256 _amount, address _tokenReceiver) public whenNotPaused nonReentrant {
+        _checkAndDeliver(_amount, _tokenReceiver);
+        // rounding up to the next whole number. Investor is charged up to one currency bit more in case of a fractional currency bit.
+        uint256 currencyAmount = calculateCurrencyAmountFromTokenAmount(_amount);
         IFeeSettingsV2 feeSettings = token.feeSettings();
         uint256 fee = feeSettings.publicFundraisingFee(currencyAmount);
         if (fee != 0) {
             currency.safeTransferFrom(_msgSender(), feeSettings.publicFundraisingFeeCollector(), fee);
         }
 
         currency.safeTransferFrom(_msgSender(), currencyReceiver, currencyAmount - fee);
-
-        token.mint(_tokenReceiver, _amount);
```

Comment:
General security advice:

the mint always happens in the end. In case there is a reentrancy attack or another not-yet-known attack, it is best practice to give out the reward in the very end. So it should stay there.

Reply:
True. I prioritized cheap reverts in case of a failing check too high. Moved it to the end 667ed3143371af1d73cf78783bfd7352e62920d8